File: scripts/download-nfts.py
"""
Download all NFT images from OpenSea CDN and save as .avif locally.
No conversion needed — modern browsers decode AVIF natively.

Run from workspace root:
  python3 scripts/download-nfts.py
"""

import json, os, sys, urllib.request, concurrent.futures, time

META  = "artifacts/museum-3d/public/metadata.json"
OUTDIR = "artifacts/museum-3d/public/nft-images"
WORKERS = 40

os.makedirs(OUTDIR, exist_ok=True)

with open(META) as f:
    meta = json.load(f)

total = len(meta)
done = skip = errors = 0
start = time.time()

def download_one(entry):
    token_id = entry["token_id"]
    url      = entry["image"]
    out_path = os.path.join(OUTDIR, f"{token_id}.avif")

    if os.path.exists(out_path) and os.path.getsize(out_path) > 1000:
        return "skip"
    try:
        req = urllib.request.Request(url, headers={
            "User-Agent": "Mozilla/5.0",
            "Accept": "image/avif,image/*",
        })
        with urllib.request.urlopen(req, timeout=20) as r:
            data = r.read()
        with open(out_path, "wb") as f:
            f.write(data)
        return "ok"
    except Exception as e:
        return f"err:{str(e)[:80]}"

with concurrent.futures.ThreadPoolExecutor(max_workers=WORKERS) as ex:
    for i, result in enumerate(ex.map(download_one, meta), 1):
        if result == "skip":   skip   += 1
        elif result == "ok":   done   += 1
        else:                  errors += 1

        if i % 200 == 0 or i == total:
            elapsed = time.time() - start
            rate    = i / elapsed
            eta     = (total - i) / rate if rate else 0
            sys.stdout.write(
                f"\r[{i}/{total}] new={done} skip={skip} err={errors}"
                f"  {rate:.0f}/s  ETA {eta:.0f}s   "
            )
            sys.stdout.flush()

print(f"\n\nDone: {done} downloaded, {skip} already existed, {errors} errors")
print(f"Total time: {time.time()-start:.0f}s")

File: scripts/download-nfts.mjs
/**
 * Downloads all 3333 NFT images from OpenSea CDN, converts AVIF → JPEG,
 * and saves them to artifacts/museum-3d/public/nft-images/{token_id}.jpg
 *
 * Run from workspace root:
 *   node scripts/download-nfts.mjs
 */

import { readFileSync, existsSync, mkdirSync, writeFileSync, unlinkSync } from "fs";
import { execFile } from "child_process";
import { promisify } from "util";
import { tmpdir } from "os";
import { join } from "path";

const execFileAsync = promisify(execFile);

const OUT_DIR   = "artifacts/museum-3d/public/nft-images";
const META_PATH = "artifacts/museum-3d/public/metadata.json";
const CONCURRENCY = 30;
const QUALITY     = "5"; // ffmpeg JPEG quality scale (2=best, 31=worst; 5 ≈ 85%)

mkdirSync(OUT_DIR, { recursive: true });

const meta  = JSON.parse(readFileSync(META_PATH, "utf8"));
const total = meta.length;

let done = 0, skipped = 0, errors = 0;

function progress() {
  const pct = ((done / total) * 100).toFixed(1);
  process.stdout.write(`\r[${pct}%] ${done}/${total}  skipped=${skipped}  errors=${errors}   `);
}

async function downloadOne({ token_id, image: url }) {
  const outPath = join(OUT_DIR, `${token_id}.jpg`);

  if (existsSync(outPath)) {
    skipped++;
    done++;
    progress();
    return;
  }

  const tmpPath = join(tmpdir(), `nft_${token_id}_${process.pid}.avif`);

  try {
    // 1. Fetch raw image bytes from CDN
    const resp = await fetch(url, {
      headers: { Accept: "image/avif, image/jpeg, image/png, image/*" },
    });
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const bytes = Buffer.from(await resp.arrayBuffer());
    writeFileSync(tmpPath, bytes);

    // 2. Convert to JPEG with ffmpeg
    await execFileAsync("ffmpeg", [
      "-y",
      "-i", tmpPath,
      "-q:v", QUALITY,
      "-vf", "scale=512:-1",   // resize to max 512px wide — reduces file size
      outPath,
    ], { timeout: 30_000 });

    done++;
  } catch (err) {
    errors++;
    done++;
    // Leave outPath absent so ProximityTextureManager falls back to CDN
    if (errors <= 20) {
      console.error(`\n  ✗ ${token_id}: ${err.message?.slice(0, 120)}`);
    }
  } finally {
    try { unlinkSync(tmpPath); } catch { /* ignore */ }
    progress();
  }
}

async function worker(queue) {
  while (queue.length > 0) {
    const entry = queue.shift();
    if (entry) await downloadOne(entry);
  }
}

console.log(`Downloading ${total} NFT images  (${CONCURRENCY} workers, quality=${QUALITY})\n`);
const queue = [...meta];
await Promise.all(Array.from({ length: CONCURRENCY }, () => worker(queue)));

console.log(`\n\nDone!  ${done - errors - skipped} new,  ${skipped} already existed,  ${errors} failed`);
if (errors > 0) {
  console.log("Failed images will fall back to CDN at runtime.");
}
